python/pyarrow/src/arrow/python/datetime.h

```diff
@@ -144,6 +144,16 @@ inline TimePoint TimePoint_from_ns(int64_t val) {
   return TimePoint(TimePoint::duration(val));
 }
 
+ARROW_PYTHON_EXPORT
+inline std::chrono::system_clock::time_point TimePoint_to_system_time(TimePoint val) {
+  return std::chrono::time_point_cast<std::chrono::system_clock::duration>(val);
```

Comment:
Just to understand this code: what does this exactly do? Because my reading is that `TimePoint` type is also already an alias for some `time_point` ?

Reply:
I see a comment in the tests about `system_clock::duration` being different depending on the platform, while `TimePoint` here is defined to always be nanoseconds? 

So for the flight API we need to system-dependent resolution?

If that's the reason for adding those APIs, could you add a comment briefly explaining what those functions do and why we need them?

Reply:
The underlying issue is `FlightEndpoint.expiration_time` having different types in C++ and Cython:
- In C++, it is `std::chrono::system_clock::time_point` (OS-dependent precision).
- In Cython, it is `std::chrono::time_point<std::chrono::system_clock, std::chrono::nanoseconds>` (OS-independent precision).

The latter is the desired type.

The [right thing to do is to fix](https://github.com/apache/arrow/pull/43537#discussion_r1751365590) `FlightEndpoint.expiration_time` in [types.h](https://github.com/apache/arrow/blob/f3dd298bd32f6dc38654680c49b3f1fbf97e3d5f/cpp/src/arrow/flight/types.h#L83), and all the casting and helper methods in this PR would go away. But that is a [breaking change](https://github.com/apache/arrow/pull/43537#discussion_r1751323212).

See https://github.com/apache/arrow/pull/43537#discussion_r1751117367.

I'll add more comments to the workaround and will raise a PR to fix the type and revert the workaround to manage that breaking change separately.